README.md
# GENIE3
GENIE3 (GEne Network Inference with Ensemble of trees): Inference of gene regulatory networks from gene expression data

GENIE3 is now available from [Bioconductor](https://bioconductor.org/packages/devel/bioc/html/GENIE3.html). A **tutorial** [(vignette)](https://bioconductor.org/packages/release/bioc/vignettes/GENIE3/inst/doc/GENIE3.html) is included in the package.


NEWS
VERSION 1.13.3
---------------------
    - The regulators can now be provided as a list with different regulators for each gene/feature.
    
VERSION 1.5.2
---------------------
    - Adding check for duplicated rownames (gene id/names)
    - Bugfix in getLinkList: diag <- NA

VERSION 0.99.7
----------------------
    - Solved error for not found "%dopar%"
    
VERSION 0.99.4
----------------------
    Removed argument "seed", use set.seed instead
    Minor improvemtents
    
VERSION 0.99.2
----------------------
INTERFACE CHANGES
    - Functions and parameters renamed to CamelCase rather than containing dots (e.g. get.link.list into getLinkList)
    
NEW FEATURES
    - GENIE3 now accepts ExpressionSet, SummarizedExperiment and SCESet as input
    

tests/testthat.R

testthat::test_check("GENIE3")


tests/testthat/test_GENIE3.R
.checkGenie3_output <- function(weightMatrix, exprMatrix)
{
  testthat::expect_equal(nrow(weightMatrix), 20)
  testthat::expect_equal(ncol(weightMatrix), 20)
  testthat::expect_true(is.numeric(weightMatrix))
  testthat::expect_equal(sum(is.na(weightMatrix)), 0)
  testthat::expect_equal(sum(weightMatrix < 0), 0)
  testthat::expect_equal(sum(diag(weightMatrix)), 0)
  testthat::expect_equal(sum(rownames(weightMatrix) == colnames(weightMatrix)), 20)
  testthat::expect_equal(sum(rownames(weightMatrix) == rownames(exprMatrix)), 20)
}

test_that("GENIE3 tests", {

	exprMatrix <- matrix(sample(1:10, 100, replace=TRUE), nrow=20)
	rownames(exprMatrix) <- paste("Gene", 1:20, sep="")
	colnames(exprMatrix) <- paste("Sample", 1:5, sep="")

	# Default parameters
	set.seed(123)
	weightMatrix <- GENIE3(exprMatrix, nCores=1)

	.checkGenie3_output(weightMatrix, exprMatrix)

  #### Regulators as number
	regulators <- c(2,4,6)
	weightMatrix <- GENIE3(exprMatrix, nTree=100, regulators=regulators)
	testthat::expect_equal(sum(weightMatrix < 0), 0)
	testthat::expect_equal(nrow(weightMatrix), 3)
	testthat::expect_equal(ncol(weightMatrix), 20)
	testthat::expect_true(is.numeric(weightMatrix))
	testthat::expect_equal(sum(is.na(weightMatrix)), 0)
	testthat::expect_equal(sum(diag(weightMatrix[,rownames(weightMatrix)])), 0)
	# testthat::expect_equal(sum(rownames(weightMatrix) == colnames(weightMatrix)), 20)
	testthat::expect_equal(sum(colnames(weightMatrix) == rownames(exprMatrix)), 20)

	#### Regulators as name
	regulators <- rownames(exprMatrix)[c(5,9)]
	weightMatrix <- GENIE3(exprMatrix, treeMethod="ET", K=1, nTree=100, regulators=regulators)
	testthat::expect_equal(sum(weightMatrix < 0), 0)
	testthat::expect_equal(nrow(weightMatrix), 2)
	testthat::expect_equal(ncol(weightMatrix), 20)
	testthat::expect_true(is.numeric(weightMatrix))
	testthat::expect_equal(sum(is.na(weightMatrix)), 0)
	testthat::expect_equal(sum(diag(weightMatrix[,rownames(weightMatrix)])), 0)
	# testthat::expect_equal(sum(rownames(weightMatrix) == colnames(weightMatrix)), 20)
	testthat::expect_equal(sum(colnames(weightMatrix) == rownames(exprMatrix)), 20)

	#### Targets as number
	regulators <- c(2,4,6)
	targets <- c(1,5)
	weightMatrix <- GENIE3(exprMatrix, regulators=regulators, targets=targets)
	testthat::expect_equal(nrow(weightMatrix), 3)
	testthat::expect_equal(ncol(weightMatrix), 2)

	#### Targets as name
	targets <- rownames(exprMatrix)[c(5,9)]
	weightMatrix <- GENIE3(exprMatrix, targets=targets)
	testthat::expect_equal(nrow(weightMatrix), 20)
	testthat::expect_equal(ncol(weightMatrix), 2)

	#### Only one target (Note that the regulator is included, and will take all the weight...)
	weightMatrix <- GENIE3(exprMatrix, regulators=2:3, targets=3)
	testthat::expect_equal(ncol(weightMatrix), 1)
	
	#### Different regulators for each gene & return as list
	regulatorsList <- list("Gene1"=rownames(exprMatrix)[1:10],
	                       "Gene2"=rownames(exprMatrix)[10:20],
	                       "Gene20"=rownames(exprMatrix)[15:20])

	set.seed(123)
	weightList <- GENIE3(exprMatrix, nCores=1, targets=names(regulatorsList), regulators=regulatorsList, returnMatrix=FALSE)
	testthat::expect_equal(lengths(weightList), setNames(c(9,11,5),names(regulatorsList)))

	### Other input classes:
	eset <- Biobase::ExpressionSet(assayData=exprMatrix)
	testthat::expect_equivalent(class(GENIE3(eset))[1], "matrix")

	sexp <- SummarizedExperiment::SummarizedExperiment(assays=list(counts=exprMatrix))
	wm <- GENIE3(sexp)
	testthat::expect_equivalent(class(wm)[1], "matrix")

	# sce <- scater::newSCESet(countData=exprMatrix)
	# testthat::expect_equal(class(GENIE3(sce)), "matrix")

	### Multicore
	# set.seed(123)
	# weightMatrix_multicore1 <- GENIE3(exprMatrix, nCores=4)
	# set.seed(123)
	# weightMatrix_multicore2 <- GENIE3(exprMatrix, nCores=4)
	# testthat::expect_equal(weightMatrix_multicore1, weightMatrix_multicore2)
	# 
	# .checkGenie3_output(weightMatrix_multicore1, exprMatrix)
	# set.seed(123)
	# weightMat <- GENIE3(exprMatr, nCores=10, targets=names(regulatorsList), regulators=regulatorsList, returnMatrix=FALSE); weightMat
	# 
})



tests/testthat/test_tlGENIE3.R
test_that("tlGENIE3 works properly", {
	
	data1 <- matrix(sample(1:10, 70, replace=TRUE), nrow=10)
	rownames(data1) <- paste("Gene", 1:10, sep="")
	colnames(data1) <- paste("Time point", 1:7, sep="")
	
	data2 <- matrix(sample(1:10, 50, replace=TRUE), nrow=10)
	rownames(data2) <- paste("Gene", 1:10, sep="")
	colnames(data2) <- paste("Time point", 1:5, sep="")
	
	TS.data <- list(data1,data2)
	
	SS.data <- matrix(sample(1:10, 100, replace=TRUE), nrow=10)
	rownames(SS.data) <- paste("Gene", 1:10, sep="")
	colnames(SS.data) <- paste("Sample", 1:10, sep="")
	
	
	weight.matrix <- tlGENIE3(TS.data)
	
	expect_equal(dim(weight.matrix)[1], 10)
	expect_equal(dim(weight.matrix)[2], 10)
	expect_true(is.numeric(weight.matrix))
	expect_equal(sum(is.na(weight.matrix)), 0)
	expect_equal(sum(weight.matrix < 0), 0)
	expect_equal(sum(diag(weight.matrix)), 0)
	expect_equal(sum(rownames(weight.matrix) == colnames(weight.matrix)), 10)
	expect_equal(sum(rownames(weight.matrix) == rownames(TS.data[[1]])), 10)
	
	
	weight.matrix <- tlGENIE3(TS.data,SS.data=NULL)
	
	expect_equal(dim(weight.matrix)[1], 10)
	expect_equal(dim(weight.matrix)[2], 10)
	expect_true(is.numeric(weight.matrix))
	expect_equal(sum(is.na(weight.matrix)), 0)
	expect_equal(sum(weight.matrix < 0), 0)
	expect_equal(sum(diag(weight.matrix)), 0)
	expect_equal(sum(rownames(weight.matrix) == colnames(weight.matrix)), 10)
	expect_equal(sum(rownames(weight.matrix) == rownames(TS.data[[1]])), 10)
	
	
	weight.matrix <- tlGENIE3(TS.data,SS.data=SS.data)
	
	expect_equal(dim(weight.matrix)[1], 10)
	expect_equal(dim(weight.matrix)[2], 10)
	expect_true(is.numeric(weight.matrix))
	expect_equal(sum(is.na(weight.matrix)), 0)
	expect_equal(sum(weight.matrix < 0), 0)
	expect_equal(sum(diag(weight.matrix)), 0)
	expect_equal(sum(rownames(weight.matrix) == colnames(weight.matrix)), 10)
	expect_equal(sum(rownames(weight.matrix) == rownames(TS.data[[1]])), 10)
	
	
	regulators <- c(2,4,6)
	weight.matrix <- tlGENIE3(TS.data, h=3, tree.method="ET", K="all", ntrees=100, regulators=regulators)
	zidx <- setdiff(1:10, regulators)
	expect_equal(sum(weight.matrix[zidx,]), 0)
	expect_equal(sum(weight.matrix < 0), 0)
	expect_equal(dim(weight.matrix)[1], 10)
	expect_equal(dim(weight.matrix)[2], 10)
	expect_true(is.numeric(weight.matrix))
	expect_equal(sum(is.na(weight.matrix)), 0)
	expect_equal(sum(diag(weight.matrix)), 0)
	expect_equal(sum(rownames(weight.matrix) == colnames(weight.matrix)), 10)
	expect_equal(sum(rownames(weight.matrix) == rownames(TS.data[[1]])), 10)
	
	
	regulators <- c("Gene5","Gene9")
	weight.matrix <- tlGENIE3(TS.data, tree.method="ET", K=1, ntrees=100, regulators=regulators)
	zidx <- setdiff(rownames(weight.matrix), regulators)
	expect_equal(sum(weight.matrix[zidx,]), 0)
	expect_equal(sum(weight.matrix < 0), 0)
	expect_equal(dim(weight.matrix)[1], 10)
	expect_equal(dim(weight.matrix)[2], 10)
	expect_true(is.numeric(weight.matrix))
	expect_equal(sum(is.na(weight.matrix)), 0)
	expect_equal(sum(diag(weight.matrix)), 0)
	expect_equal(sum(rownames(weight.matrix) == colnames(weight.matrix)), 10)
	expect_equal(sum(rownames(weight.matrix) == rownames(TS.data[[1]])), 10)

})


R/getLinkList.R
#' @title getLinkList
#'
#' @description \code{getLinkList} Converts the weight matrix, as returned by \code{\link{GENIE3}}, to a sorted list of regulatory links (most likely links first).
#'
#' @param weightMatrix Weighted adjacency matrix as returned by \code{\link{GENIE3}}.
#' @param reportMax Maximum number of links to report. The default value NULL means that all the links are reported.
#' @param threshold Only links with a weight equal or above the threshold are reported. Default: threshold = 0, i.e. all the links are reported.
#'
#' @return List of regulatory links in a data frame. Each line of the data frame corresponds to a link. The first column is the regulatory gene, the second column is the target gene, and the third column is the weight of the link.
#'
#' @seealso \code{\link{GENIE3}}
#'
#' @examples
#' ## Generate fake expression matrix
#' exprMat <- matrix(sample(1:10, 100, replace=TRUE), nrow=20)
#' rownames(exprMat) <- paste("Gene", 1:20, sep="")
#' colnames(exprMat) <- paste("Sample", 1:5, sep="")
#'
#' ## Run GENIE3
#' weightMat <- GENIE3(exprMat, regulators=paste("Gene", 1:5, sep=""))
#'
#' ## Get ranking of edges
#' linkList <- getLinkList(weightMat)
#' head(linkList)
#' @export
getLinkList <- function(weightMatrix, reportMax=NULL, threshold=0) {
    if(!is.numeric(threshold)) {
    	stop("threshold must be a number.")
    }

	# Only process weights off-diagonal
  regulatorsInTargets <- rownames(weightMatrix)[rownames(weightMatrix) %in% colnames(weightMatrix)]
  if(length(regulatorsInTargets)==1) weightMatrix[regulatorsInTargets, regulatorsInTargets] <- NA
  if(length(regulatorsInTargets)>1) diag(weightMatrix[regulatorsInTargets, regulatorsInTargets]) <- NA

  linkList <- reshape2::melt(weightMatrix, na.rm=TRUE)
  colnames(linkList) <- c("regulatoryGene", "targetGene", "weight")
  linkList <- linkList[linkList$weight>=threshold,]
  linkList <- linkList[order(linkList$weight, decreasing=TRUE),]
  
  if(!is.null(reportMax)) {
  	linkList <- linkList[1:min(nrow(linkList), reportMax),]
  }
  rownames(linkList) <- NULL
  
  uniquePairs <- nrow(unique(linkList[,c("regulatoryGene", "targetGene")]))
  if(uniquePairs < nrow(linkList)) 
    warning("There might be duplicated regulator-target (gene id/name) pairs.")
  
  return(linkList)
}


R/get.link.list.R
#' @title get.link.list
#'
#' @description \code{get.link.list} Converts the weight matrix, as returned by \code{\link{GENIE3}}, to a sorted list of regulatory links (most likely links first).
#'
#' @param weight.matrix Weighted adjacency matrix as returned by \code{\link{GENIE3}}.
#' @param report.max Maximum number of links to report. The default value NULL means that all the links are reported.
#' @param threshold Only links with a weight above the threshold are reported. Default: threshold = 0, i.e. all the links are reported.
#'
#' @return List of regulatory links in a data frame. Each line of the data frame corresponds to a link. The first column is the regulatory gene, the second column is the target gene, and the third column is the weight of the link.
#'
#' @seealso \code{\link{GENIE3}}
#'
#' @examples
#' ## Generate fake expression matrix
#' expr.matrix <- matrix(sample(1:10, 100, replace=TRUE), nrow=20)
#' rownames(expr.matrix) <- paste("Gene", 1:20, sep="")
#' colnames(expr.matrix) <- paste("Sample", 1:5, sep="")
#'
#' ## Run GENIE3
#' weight.matrix <- GENIE3(expr.matrix, regulators=paste("Gene", 1:5, sep=""))
#'
#' ## Get ranking of edges
#' link.list <- get.link.list(weight.matrix)
#' head(link.list)
#' @export
get.link.list <- function(weight.matrix, report.max=NULL, threshold=0) {
    if(!is.numeric(threshold)) {
    	stop("threshold must be a number.")
    }

	# Only process weights off-diagonal
	diag(weight.matrix) <- NA
    link.list <- reshape2::melt(weight.matrix, na.rm=TRUE)
    colnames(link.list) <- c("regulatory.gene", "target.gene", "weight")
    link.list <- link.list[link.list$weight>=threshold,]
    link.list <- link.list[order(link.list$weight, decreasing=TRUE),]

    if(!is.null(report.max)) {
    	link.list <- link.list[1:min(nrow(link.list), report.max),]
    }

    rownames(link.list) <- NULL

    return(link.list)
}


R/GENIE3.R
#' @title GENIE3
#'
#' @description \code{GENIE3} Infers a gene regulatory network (in the form of a weighted adjacency matrix) from expression data, using ensembles of regression trees.
#'
#' @param exprMatrix Expression matrix (genes x samples). Every row is a gene, every column is a sample.
#' The expression matrix can also be provided as one of the Bioconductor classes:
#' \itemize{
#' \item \code{ExpressionSet}: The matrix will be obtained through exprs(exprMatrix)
#' \item \code{RangedSummarizedExperiment}: The matrix will be obtained through assay(exprMatrix), wich will extract the first assay (usually the counts)
#' }
#' @param regulators Subset of genes used as candidate regulators. Must be either a vector of gene names, e.g. \code{c("at_12377", "at_10912")} or a vector of indices, e.g. \code{c(1,5,6,7)}. 
#' The default value NULL means that all the genes are used as candidate regulators (which is NOT recommended). To provide different regulators for each gene, providethem as named list.
#' @param targets Subset of genes to which potential regulators will be calculated. Must be either a vector of indices, e.g. \code{c(1,5,6,7)}, or a vector of gene names, e.g. \code{c("at_12377", "at_10912")}. If NULL (default), regulators will be calculated for all genes in the input matrix.
#' @param treeMethod Tree-based method used. Must be either "RF" for Random Forests (default) or "ET" for Extra-Trees.
#' @param K Number of candidate regulators randomly selected at each tree node (for the determination of the best split). Must be either "sqrt" for the square root of the total number of candidate regulators (default), "all" for the total number of candidate regulators, or a stricly positive integer.
#' @param nTrees Number of trees in an ensemble for each target gene. Default: 1000.
#' @param nCores Number of cores to use for parallel computing. Default: 1.
#' @param returnMatrix Returns output as weight matrix (TRUE). Otherwise (FALSE) it is returned as a list.
#' @param verbose If set to TRUE, a feedback on the progress of the calculations is given. Default: FALSE.
#'
#' @return Weighted adjacency matrix of inferred network. Element w_ij (row i, column j) gives the importance of the link from regulatory gene i to target gene j.
#'
#' @examples
#' ## Generate fake expression matrix
#' exprMatrix <- matrix(sample(1:10, 100, replace=TRUE), nrow=20)
#' rownames(exprMatrix) <- paste("Gene", 1:20, sep="")
#' colnames(exprMatrix) <- paste("Sample", 1:5, sep="")
#' 
#' ## Run GENIE3
#' set.seed(123) # For reproducibility of results
#' weightMatrix <- GENIE3(exprMatrix, regulators=paste("Gene", 1:5, sep=""))
#' 
#' ## Get ranking of edges
#' linkList <- getLinkList(weightMatrix)
#' head(linkList)
#' 
#' ## Different regulators for each gene & return as list
#' regulatorsList <- list("Gene1"=rownames(exprMatrix)[1:10],
#'                        "Gene2"=rownames(exprMatrix)[10:20],
#'                        "Gene20"=rownames(exprMatrix)[15:20])
#' 
#' set.seed(123)
#' weightList <- GENIE3(exprMatrix, nCores=1, targets=names(regulatorsList), regulators=regulatorsList, returnMatrix=FALSE)
#'
#' @export
setGeneric("GENIE3", signature="exprMatrix",
function(exprMatrix, regulators=NULL, targets=NULL, 
         treeMethod="RF", K="sqrt", nTrees=1000, 
         nCores=1, returnMatrix=TRUE, verbose=FALSE)
{
  standardGeneric("GENIE3")
})

#' @export
setMethod("GENIE3", "matrix",
function(exprMatrix, regulators=NULL, targets=NULL, treeMethod="RF", K="sqrt", nTrees=1000, nCores=1, returnMatrix=TRUE, verbose=FALSE)
{
  .GENIE3(exprMatrix=exprMatrix, regulators=regulators, targets=targets, treeMethod=treeMethod, K=K, nTrees=nTrees, nCores=nCores, returnMatrix=returnMatrix, verbose=verbose)
})

#' @export
setMethod("GENIE3", "SummarizedExperiment",
function(exprMatrix, regulators=NULL, targets=NULL, treeMethod="RF", K="sqrt", nTrees=1000, nCores=1, returnMatrix=TRUE, verbose=FALSE)
{
  if(length(SummarizedExperiment::assays(exprMatrix))>1) warning("More than 1 assays are available. Only using the first one.")
  exprMatrix <- SummarizedExperiment::assay(exprMatrix)
  .GENIE3(exprMatrix=exprMatrix, regulators=regulators, targets=targets, treeMethod=treeMethod, K=K, nTrees=nTrees, nCores=nCores, returnMatrix=returnMatrix, verbose=verbose)
})

#' @export
setMethod("GENIE3", "ExpressionSet",
function(exprMatrix, regulators=NULL, targets=NULL, treeMethod="RF", K="sqrt", nTrees=1000, nCores=1, returnMatrix=TRUE, verbose=FALSE)
{
  exprMatrix <- Biobase::exprs(exprMatrix)
  .GENIE3(exprMatrix=exprMatrix, regulators=regulators, targets=targets, treeMethod=treeMethod, K=K, nTrees=nTrees, nCores=nCores, returnMatrix=returnMatrix, verbose=verbose)
})

.GENIE3 <- function(exprMatrix, regulators, targets, treeMethod, K, nTrees, nCores, returnMatrix, verbose)
{
  .checkArguments(exprMatrix=exprMatrix, regulators=regulators, targets=targets, treeMethod=treeMethod, K=K, nTrees=nTrees, nCores=nCores, verbose=verbose)
  
  ############################################################
  # transpose expression matrix to (samples x genes)
  exprMatrixT <- t(exprMatrix); rm(exprMatrix)
  num.samples <- nrow(exprMatrixT)
  allGeneNames <- colnames(exprMatrixT)

  # Regulators
  regulatorsList <- NULL
  if(is.list(regulators)) {
    regulatorsList <- regulators
    regulators <- unique(unlist(regulators))
  }
  if(is.null(regulators))
  {
    regulatorNames <- allGeneNames
  } else
  {
    # Input genes given as names
    regulatorNames <- regulators
    
    # Input genes given as indices -> get names
    if (is.numeric(regulators))
    {
      regulatorNames <- allGeneNames[regulators]
    } 
    # for security, abort if some input gene name is not in gene names
    missingGeneNames <- setdiff(regulatorNames, allGeneNames)
    if (length(missingGeneNames) != 0) stop(paste("Regulator genes missing from the expression matrix:", paste(missingGeneNames, collapse=", ")))
  }
  regulatorNames <- regulatorNames
  rm(regulators)

  # Target genes
  if(is.null(targets))
  {
    targetNames <- allGeneNames
  } else
  {
    targetNames <- targets
    
    # input gene indices given as integers
    if (is.numeric(targets))
    {
      targetNames <- allGeneNames[targets]
      
    } 
    #  abort if some input gene name is not in gene names
    missingGeneNames <- setdiff(targetNames, allGeneNames)
    if (length(missingGeneNames) != 0) stop(paste("Target genes missing from the expression matrix:", paste(missingGeneNames, collapse=", ")))
  }
  targetNames <- targetNames
  nGenes <- length(targetNames)
  rm(targets)

  # tree method
  if (treeMethod == 'RF')
  {
  	RF_randomisation <- 1
  	ET_randomisation <- 0
  	bootstrap_sampling <- 1
  } else {
  	RF_randomisation <- 0
  	ET_randomisation <- 1
  	bootstrap_sampling <- 0
  }

  if (verbose) message(paste("Tree method: ", treeMethod,
                             "\nK: ", K,
                              "\nNumber of trees: ", nTrees, sep=""))
  # other default parameters
  nmin <- 1
  permutation_importance <- 0
  
  # compute importances for every target gene
  nCores <- min(nCores, length(targetNames))
  if(nCores==1)
  {
    # serial computing
    if(verbose) message("Using 1 core.")
    
    ret<-list()
    if(returnMatrix)
    {
      # setup weight matrix
      weightMatrix <- matrix(0.0, nrow=length(regulatorNames), ncol=length(targetNames))
      rownames(weightMatrix) <- regulatorNames
      colnames(weightMatrix) <- targetNames
    }
    
    for(targetName in targetNames)
    {
      if(verbose) message(paste("Computing gene ", which(targetNames == targetName), "/", nGenes, ": ",targetName, sep=""))

      # remove target gene from input genes      
      theseRegulatorNames <- regulatorNames
      if(!is.null(regulatorsList)) theseRegulatorNames <- regulatorsList[[targetName]]
      theseRegulatorNames <- setdiff(theseRegulatorNames, targetName)
        
      numRegulators <- length(theseRegulatorNames)
      mtry <- .setMtry(K, numRegulators)
    
      x <- exprMatrixT[,theseRegulatorNames]
      y <- exprMatrixT[,targetName]
    
      im <- .C("BuildTreeEns",as.integer(num.samples),as.integer(numRegulators),
            as.single(c(x)),as.single(c(y)),as.integer(nmin),
            as.integer(ET_randomisation),as.integer(RF_randomisation),
            as.integer(mtry),as.integer(nTrees),
            as.integer(bootstrap_sampling),as.integer(permutation_importance),
            as.double(vector("double",numRegulators)))[[12]]
    
      # normalize variable importances
      im <- im / sum(im)
    
      if(returnMatrix) {
        weightMatrix[theseRegulatorNames, targetName] <- im
      }else{
        ret[[targetName]] <- setNames(im, theseRegulatorNames)
      }
    }
    if(returnMatrix) ret <- weightMatrix
   
  } else
  {
      doParallel::registerDoParallel(cores=nCores)
      if(verbose) message(paste("\nUsing", foreach::getDoParWorkers(), "cores."))

      "%dopar%"<- foreach::"%dopar%"
      suppressPackageStartupMessages(ret <- doRNG::"%dorng%"(foreach::foreach(targetName=targetNames),
      {
          # remove target gene from input genes      
          theseRegulatorNames <- regulatorNames
          if(!is.null(regulatorsList)) theseRegulatorNames <- regulatorsList[[targetName]]
          theseRegulatorNames <- setdiff(theseRegulatorNames, targetName)
          
          numRegulators <- length(theseRegulatorNames)
          mtry <- .setMtry(K, numRegulators)

          x <- exprMatrixT[,theseRegulatorNames]
          y <- exprMatrixT[,targetName]

          im <- .C("BuildTreeEns", as.integer(num.samples), as.integer(numRegulators),
          as.single(c(x)),as.single(c(y)), as.integer(nmin),
          as.integer(ET_randomisation), as.integer(RF_randomisation),
          as.integer(mtry), as.integer(nTrees),
          as.integer(bootstrap_sampling), as.integer(permutation_importance),
          as.double(vector("double",numRegulators)))[[12]]

          # normalize variable importances
          im <- im / sum(im)

          setNames(list(setNames(im, theseRegulatorNames)), targetName)
      }))
      attr(ret, "rng") <- NULL
      attr(ret,"doRNG_version") <- NULL
      ret <- unlist(ret, recursive=FALSE)
      
      if(returnMatrix)
      {
        weightMatrix <- dplyr::bind_rows(ret)
        weightMatrix <- as.matrix(weightMatrix)
        rownames(weightMatrix) <- names(ret)
        rm(ret)
        weightMatrix <- t(weightMatrix)
        weightMatrix[which(is.na(weightMatrix), arr.ind = TRUE)] <- 0 # optional?  
        ret <- weightMatrix
      }
      # else{ # as linkList instead
      ## needs to skip # ret <- unlist(ret, recursive=F)
      #   ret <- lapply(ret, reshape2::melt)
      #   ret <- lapply(ret, function(x) data.frame(regulator=rownames(x), x))
      #   ret <- as.data.frame(data.table::rbindlist(ret))
      #   colnames(ret) <- c("regulatoryGene", "weight", "targetGene")
      #   ret <- ret[,c("regulatoryGene", "targetGene", "weight")]
      #   ret <- ret[ret$weight>0,,drop=F]
      #   ret <- ret[order(-ret$weight),]
      #   rownames(ret) <- NULL
      # }
  }
  
  return(ret)
}

# mtry <- setMtry(K, numRegulators)
.setMtry <- function(K, numRegulators)
{
  # set mtry
  if (class(K) == "numeric") {
    mtry <- K
  } else if (K == "sqrt") {
    mtry <- round(sqrt(numRegulators))
  } else {
    mtry <- numRegulators
  }

  return(mtry)
}

.checkArguments <- function(exprMatrix, regulators, targets, treeMethod, K, nTrees, nCores, verbose)
{
  ############################################################
  # check input arguments
  if (!is.matrix(exprMatrix) && !is.array(exprMatrix)) {
    stop("Parameter exprMatrix must be a two-dimensional matrix where each row corresponds to a gene and each column corresponds to a condition/sample/cell.")
  }
  
  if (length(dim(exprMatrix)) != 2) {
    stop("Parameter exprMatrix must be a two-dimensional matrix where each row corresponds to a gene and each column corresponds to a condition/sample/cell.")
  }
  
  if (is.null(rownames(exprMatrix))) {
    stop("exprMatrix must contain the names of the genes as rownames.")
  }
  
  countGeneNames <- table(rownames(exprMatrix))
  nonUniqueGeneNames <- countGeneNames[countGeneNames>1]
  if(length(nonUniqueGeneNames)>0) 
    stop("The following gene IDs (rownames) are not unique: ", paste(names(nonUniqueGeneNames), collapse=", "))
  
  if (treeMethod != "RF" && treeMethod != "ET") {
    stop("Parameter treeMethod must be \"RF\" (Random Forests) or \"ET\" (Extra-Trees).")
  }
  
  if (K != "sqrt" && K != "all" && !is.numeric(K)) {
    stop("Parameter K must be \"sqrt\", or \"all\", or a strictly positive integer.")
  }
  
  if (is.numeric(K) && K<1) {
    stop("Parameter K must be \"sqrt\", or \"all\", or a strictly positive integer.")
  }
  
  if (!is.numeric(nTrees) || nTrees<1) {
    stop("Parameter nTrees should be a stricly positive integer.")
  }

  if (!is.null(targets))
  {
    if (!is.vector(targets)) {
      stop("Parameter 'targets' must a vector (of indices or gene names).")
    }

    if (is.numeric(targets)) {
      if(max(targets) > nrow(exprMatrix)) stop("At least one index in 'targets' exceeds the number of genes.")
      if(min(targets) <= 0) stop("The indexes in 'targets' should be >=1.")
    }

    if(any(table(targets)>1))
      stop("Please, provide each target (name/ID) only once.")

    if(is.character(targets)){
      targetsInMatrix <- intersect(targets, rownames(exprMatrix))
      if(length(targetsInMatrix) == 0)
        stop("The genes must contain at least one target.")

      if(length(targetsInMatrix) < length(targets))
        warning("Only", length(targetsInMatrix), "out of", length(targets), "target genes (IDs/names) are in the expression matrix.")
    }
  }

  if(is.list(regulators))
  {
    if(!all(names(regulators) %in% targets)) stop("Regulators: If provided as a named list, all names should be targets.")
    regulators <- unique(unlist(regulators))
  }
  if (!is.null(regulators))
  {
    if(length(regulators)<2) stop("Provide at least 2 potential regulators.")
    
    if (!is.vector(regulators)) {
      stop("Parameter 'regulators' must a vector (of indices or gene names).")
    }
    
    if (is.numeric(regulators)) {
      if(max(regulators) > nrow(exprMatrix)) stop("At least one index in 'regulators' exceeds the number of genes.")
      if(min(regulators) <= 0) stop("The indexes in 'regulators' should be >=1.")
    }
    
    if(any(table(regulators)>1))
      stop("Please, provide each regulator (name/ID) only once.")
    
    if (is.character(regulators)){
      regulatorsInMatrix <- intersect(regulators, rownames(exprMatrix))
      if(length(regulatorsInMatrix) <2)
        stop("Fewer than 2 regulators in the expression matrix (rownames).")
      
      if(length(regulatorsInMatrix) < length(regulators))
        warning("Only", length(regulatorsInMatrix), "out of", length(regulators), " candidate regulators (IDs/names) are in the expression matrix.")
    }
  }
  
  if (!is.numeric(nCores) || nCores<1)
  {
    stop("Parameter nCores should be a stricly positive integer.")
  }
}
